### PR TITLE
[ownership] Struct definitions and conversions for Rust and Firmware

### PR DESCRIPTION
**Pull request by cfrantz**
1. Create C structs for ownership-related configuration in firmware.
2. Create Rust versions of those structs and conversion functions for dealing with the structs in json and binary form.
3. Add opentitantool commands to convert the structs between json and binary.

### CODE REVIEW

File: sw/host/opentitanlib/src/ownership/owner.rs [hunk 1]
```diff
@@ -0,0 +1,594 @@
+// Copyright lowRISC contributors.
```
Comment (moidx) [inline sw/host/opentitanlib/src/ownership/owner.rs:1]: I just realized the copyright header may have been changed. Need to check copyright linter results.

File: sw/host/opentitanlib/src/ownership/owner.rs [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+use serde_annotate::Annotate;
+use std::convert::TryFrom;
+use std::io::{Read, Write};
+
+use super::misc::{KeyMaterial, OwnershipKeyAlg, TlvHeader, TlvTag};
+use super::{OwnerApplicationKey, OwnerFlashConfig, OwnerFlashInfoConfig, OwnerRescueConfig};
+use crate::crypto::ecdsa::{EcdsaPrivateKey, EcdsaRawSignature};
+use crate::with_unknown;
+
+with_unknown! {
```
Comment (moidx) [inline sw/host/opentitanlib/src/ownership/owner.rs:17]: What does `with_unkown!` do?

File: sw/device/silicon_creator/lib/ownership/datatypes.h [hunk 3]
```diff
@@ -41,6 +44,30 @@ typedef enum ownership_state {
   kOwnershipStateLockedNone = 0,
 } ownership_state_t;
 
+typedef enum ownership_key_alg {
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/datatypes.h:47]: Fixed.  The enum name is now `EcdsaP256` and the FourCC is `P256`.

File: sw/device/silicon_creator/lib/ownership/datatypes.h [hunk 4]
```diff
@@ -100,6 +127,155 @@ OT_ASSERT_MEMBER_OFFSET(owner_block_t, signature, 1952);
...
+   * tag: `APPK`.
+   * length: 48 + sizeof(key).
+   */
+  tlv_header_t header;
+  /** Key algorithm.  One of ECDSA, SPX+ or SPXq20. */
+  uint32_t key_alg;
+  /** Key domain.  Recognized values: PROD, DEV, TEST */
+  uint32_t key_domain;
+  /** Key diversifier. */
+  uint32_t key_diversifier[7];
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/datatypes.h:158]: Done.

File: sw/device/silicon_creator/lib/ownership/datatypes.h [hunk 5]
```diff
@@ -100,6 +127,155 @@ OT_ASSERT_MEMBER_OFFSET(owner_block_t, signature, 1952);
...
+  /** Usage constraint must match manifest header's constraint */
+  uint32_t usage_constraint;
+  /** Key material.  Varies by algorithm type. */
+  union {
+    uint32_t id;
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/device/silicon_creator/lib/ownership/datatypes.h:163]: This is using the union approach using the assumption that the first word of the public key is the identifier.

File: sw/host/opentitanlib/src/ownership/application_key.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+use std::io::{Read, Write};
+
+use super::misc::{KeyMaterial, OwnershipKeyAlg, TlvHeader, TlvTag};
+use crate::with_unknown;
+
+with_unknown! {
+    pub enum ApplicationKeyDomain: u32 [default = Self::Unknown] {
+        Unknown = 0,
+        Test = u32::from_le_bytes(*b"test"),
+        Dev = u32::from_le_bytes(*b"dev\0"),
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/ownership/application_key.rs:19]: Yes. Good catch.

File: sw/host/opentitanlib/src/ownership/flash.rs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+            program: true,
+            erase: true,
+            scramble: true,
+            ecc: true,
+            high_endurance: true,
+            ..Default::default()
+        }
+    }
+
+    pub fn rom_ext() -> Self {
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/ownership/flash.rs:55]: They are suggested flags per-region type used by the "basic" owner config.  Added doc-comments.

File: sw/host/opentitanlib/src/ownership/flash.rs [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+impl OwnerFlashConfig {
+    const BASE_SIZE: usize = 8;
+    pub fn basic() -> Self {
+        Self {
+            header: TlvHeader::new(TlvTag::FlashConfig, 0),
+            config: vec![
+                OwnerFlashRegion::new(0, 32, FlashFlags::rom_ext()),
+                OwnerFlashRegion::new(32, 384, FlashFlags::firmware()),
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/ownership/flash.rs:188]: Good catch.  Fixed.

File: sw/host/opentitanlib/src/ownership/owner.rs (hunk 2 above)
Comment (cfrantz) [reply sw/host/opentitanlib/src/ownership/owner.rs:17]: With unknown is a utility macro for creating so-called Open Enums.

Rust enums are closed enums, meaning all un-enumerated values are not legal values.  C enums, OTOH, are open enums, meaning the enum _may_ take on values not enumerated in the definition.

The various `with_unknown` enums here are user-input, and thus might have values "unknown" to the enum definition (they _shouldn't_, but they _might_).

In order to accomplish this in Rust, the `with_unknown` macro rewrites `enum Foo { ... }` as `struct Foo(u32)` and then creates struct-level consts for all of the enumerators.  It also creates a number of trait implementations, including string conversions, Display, Debug, and serde Serialize/Deserialize.  These implementations make the enum useful in several contexts in opentitantool/lib (namely: CLI arguments and serialization to/from json).

File: sw/host/opentitanlib/src/ownership/owner.rs (hunk 1 above)
Comment (cfrantz) [reply sw/host/opentitanlib/src/ownership/owner.rs]: Fixed.

File: sw/device/silicon_creator/lib/ownership/datatypes.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/silicon_creator/lib/ownership/datatypes.h:167]: I missed the union definition. This LGTM.

File: sw/host/opentitanlib/src/ownership/owner.rs (hunk 2 above)
Comment (moidx) [reply sw/host/opentitanlib/src/ownership/owner.rs:17]: Thanks!

**Review by moidx — APPROVED** [review]

**Comment by cfrantz** [comment]
I don't believe the CI failures are related to this PR; this PR contains only struct definitions and rust conversion functions for those structs.

**Pull request merged**